### PR TITLE
Fix BN_nnmod and BN_mod_inverse incorrect result

### LINKED ISSUE #21110: BN_mod_inverse incorrect result when parameters are aliased
```c
#include <openssl/bn.h>

#define CF_CHECK_NE(expr, res) if ( (expr) == (res) ) { goto end; }

int main(void)
{
    char* str = NULL;
    BIGNUM* a = BN_new();
    BIGNUM* b = BN_new();
    BN_CTX* ctx = BN_CTX_new();
    CF_CHECK_NE(BN_dec2bn(&a, "5193817943"), 0);
    CF_CHECK_NE(BN_dec2bn(&b, "3259122431"), 0);
    CF_CHECK_NE(BN_mod_inverse(b, a, b, ctx), NULL);
    str = BN_bn2dec(b);
    printf("%s\n", str);
end:
    BN_free(a);
    BN_free(b);
    BN_CTX_free(ctx);
    OPENSSL_free(str);
    return 0;
}
```

This prints 0 but should print 2609653924. Note that in the call to `BN_mod_inverse`, the result is the same pointer as the modulus. The documentation explicitly states that "r may be the same BIGNUM as a or n.".

This also affects @davidben and @botovq.

### PR DESCRIPTION
**Pull request by bernd-edlinger**
When r aliases d the result may be wrong, since d is overwritten by BN_mod but it may still be used later.

Fixes #21110

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

File: test/bntest.c [hunk 1]
```diff
@@ -3165,6 +3165,33 @@ static int test_mod_exp2_mont(void)
...
+    int res = 0;
+    char *str = NULL;
+    BIGNUM* a = NULL;
+    BIGNUM* b = NULL;
+
+    if (!TEST_true(BN_dec2bn(&a, "5193817943")))
+        goto err;
+    if (!TEST_true(BN_dec2bn(&b, "3259122431")))
+        goto err;
+    if (!TEST_ptr_eq(BN_mod_inverse(b, a, b, ctx), b))
```
Comment (mattcaswell) [inline test/bntest.c:3179]: A comment stating that we are testing aliased parameters might be useful here

Comment (bernd-edlinger) [reply test/bntest.c:3180]: done.

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
@paulidale - please reconfirm

**Comment by bernd-edlinger** [comment]
Hmm, the test case mentioned here is not yet fixed:
https://github.com/openssl/openssl/issues/21110#issuecomment-1573948744
I am going to push a fix for that one in a moment and also the BN_mod_sub_quick as well.

**Review by paulidale — APPROVED** [review]

File: crypto/bn/bn_exp.c [hunk 2]
```diff
@@ -1362,6 +1362,22 @@ int BN_mod_exp_simple(BIGNUM *r, const BIGNUM *a, const BIGNUM *p,
     wstart = bits - 1;          /* The top bit of the window */
     wend = 0;                   /* The bottom bit of the window */
 
+    if (r == m) {
+        BIGNUM *m_dup = BN_CTX_get(ctx);
+
+        if (m_dup == NULL || BN_copy(m_dup, r) == NULL)
+            goto err;
+        m = m_dup;
```
Comment (tom-cosgrove-arm) [inline crypto/bn/bn_exp.c:1370]: How does `m_dup` get freed?

Comment (tom-cosgrove-arm) [reply crypto/bn/bn_exp.c:1370]: Ah, nevermind, it's the `ctx`

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
**OTC: Do want to allow aliasing the modulus in modular operations, or just error out instead?**

**Comment by bernd-edlinger** [comment]
Note: as far as BN_mod_exp_simple is concerned aliasing with the result with power (r==p) seems also affected.

**Comment by bernd-edlinger** [comment]
... while BN_exp has already explicit measures to handle aliasing p and r, although not explicitly documented this way:
https://github.com/openssl/openssl/blob/b8fa5be5506e43b405c9a3ecc3d65c77044777be/crypto/bn/bn_exp.c#L62

**Pull request closed without merging**